hs/database/JsonParser.hs

```diff
@@ -33,57 +33,58 @@ courseDirectory = "../../res/courses/"
 
 -- | A Lecture.
 data Lecture =
-    Lecture { extra      :: Int,
-              section    :: T.Text,
-              cap        :: Int,
-              time_str   :: T.Text,
-              time       :: [[Int]],
+    Lecture { extra :: Int,
+              section :: T.Text,
+              cap :: Int,
+              time_str :: T.Text,
+              time :: [[Int]],
               instructor :: T.Text,
-              enrol      :: Maybe Int,
-              wait       :: Maybe Int
+              enrol :: Maybe Int,
+              wait :: Maybe Int
             } deriving Show
 
 -- | A Tutorial.
 data Tutorial =
-    Tutorial { times       :: [[Int]],
+    Tutorial { tutorialSection :: Maybe T.Text,
+               times       :: [[Int]],
                timeStr     :: T.Text
              } deriving Show
 
 -- | A Session.
 data Session =
-    Session { lectures   :: [Lecture],
-              tutorials  :: [Tutorial]
+    Session { lectures :: [Lecture],
+              tutorials :: [Tutorial]
             } deriving Show
 
 -- | A Course.
 data Course =
-    Course { breadth               :: !T.Text,
-             description           :: !T.Text,
-             title                 :: !T.Text,
-             prereqString          :: Maybe T.Text,
-             f                     :: Maybe Session,
-             s                     :: Maybe Session,
-             y                     :: Maybe Session,
-             name                  :: !T.Text,
-             exclusions            :: Maybe T.Text,
-             manualTutorialEnrol   :: Maybe Bool,
-             distribution          :: !T.Text,
-             prereqs               :: Maybe [T.Text]
+    Course { breadth :: Maybe T.Text,
+             description :: Maybe T.Text,
+             title :: Maybe T.Text,
+             prereqString :: Maybe T.Text,
+             f :: Maybe Session,
+             s :: Maybe Session,
+             y :: Maybe Session,
+             name :: !T.Text,
+             exclusions :: Maybe T.Text,
+             manualTutorialEnrol :: Maybe Bool,
+             distribution :: Maybe T.Text,
+             prereqs :: Maybe Array
            } deriving Show
 
 instance FromJSON Course where
     parseJSON (Object v) =
-        Course <$> v .:  "breadth"
-               <*> v .:  "description"
-               <*> v .:  "title"
-               <*> v .:  "prereqString"
+        Course <$> v .:? "breadth"
+               <*> v .:? "description"
+               <*> v .:? "title"
+               <*> v .:? "prereqString"
                <*> v .:? "F"
                <*> v .:? "S"
                <*> v .:? "Y"
                <*> v .:  "name"
-               <*> v .:  "exclusions"
+               <*> v .:? "exclusions"
                <*> v .:? "manualTutorialEnrolment"
-               <*> v .:  "distribution"
+               <*> v .:? "distribution"
                <*> v .:? "prereqs"
     parseJSON _ = mzero
```

Comment:
ABP100Y1 is not very well formed. In order to account for this, many of the fields have been changed to `Maybe`.